src/token/cose/test_helper.rs

```diff
@@ -0,0 +1,534 @@
+        Some("direct") => Ok(Some(iana::Algorithm::Direct)),
+        None => Ok(None),
+        _ => Err(de::Error::custom("could not parse test case algorithm")),
+    }
+}
+
+fn deserialize_header<'de, D>(deserializer: D) -> Result<Option<Header>, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    let hdr_obj =
+        if let Some(hdr) = Option::<serde_json::Map<String, Value>>::deserialize(deserializer)? {
+            hdr
+        } else {
+            return Ok(None);
+        };
+
+    let mut builder = HeaderBuilder::new();
+
+    if let Some(v) = hdr_obj.get("kid").and_then(Value::as_str) {
+        builder = builder.key_id(v.to_string().into_bytes());
+    }
+    if let Some(v) = hdr_obj
+        .get("kid_hex")
+        .and_then(Value::as_str)
+        .map(hex::decode)
+    {
+        builder = builder.key_id(v.map_err(|e| {
+            de::Error::custom(format!("could not parse test case key ID hex: {e}"))
+        })?);
+    }
+
+    if let Some(alg) = string_to_algorithm::<D>(hdr_obj.get("alg").and_then(Value::as_str))? {
+        builder = builder.algorithm(alg);
+    }
+
+    builder = match hdr_obj.get("ctyp").and_then(Value::as_i64) {
+        Some(v) => {
+            let content_format = iana::CoapContentFormat::from_i64(v).ok_or(de::Error::custom(
+                "could not parse test case content format",
+            ))?;
+            builder.content_format(content_format)
+        }
+        None => builder,
+    };
+
+    Ok(Some(builder.build()))
+}
+
+fn deserialize_algorithm<'de, D>(deserializer: D) -> Result<Option<Algorithm>, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    let alg = if let Some(alg) = Option::<Value>::deserialize(deserializer)? {
+        alg
+    } else {
+        return Ok(None);
+    };
+    Ok(string_to_algorithm::<D>(alg.as_str())?.map(Algorithm::Assigned))
+}
+
+#[derive(Deserialize, Debug, Clone)]
+pub struct TestCase {
+    pub title: String,
+    pub description: Option<String>,
+    #[serde(default)]
+    pub fail: bool,
+    pub input: TestCaseInput,
+    pub intermediates: Option<TestCaseIntermediates>,
+    #[serde(default)]
+    pub output: TestCaseOutput,
+}
+
+#[derive(Deserialize, Debug, Clone, Default)]
+pub struct TestCaseFailures {
+    #[serde(rename = "ChangeTag")]
+    pub change_tag: Option<u64>,
+    #[serde(rename = "ChangeCBORTag")]
+    pub change_cbor_tag: Option<u64>,
+    #[serde(rename = "ChangeAttr")]
+    pub change_attribute: Option<HashMap<String, Value>>,
+    #[serde(
+        rename = "AddProtected",
+        deserialize_with = "deserialize_header",
+        default
+    )]
+    pub add_protected_headers: Option<Header>,
+    #[serde(
+        rename = "RemoveProtected",
+        deserialize_with = "deserialize_header",
+        default
+    )]
+    pub remove_protected_headers: Option<Header>,
+}
+
+#[derive(Deserialize, Debug, Clone)]
+pub struct TestCaseInput {
+    pub plaintext: String,
+    //#[serde(default)]
+    //pub detached: bool,
```

Comment:
I kept it in for completeness, as this is a field that is present in the JSON files, although we don't use it for our own tests yet, and leaving it in will make the linter complain about the unused field.

As it might come in handy, later, I'd just keep it in as a comment.